SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp

```diff
@@ -0,0 +1,370 @@
+
+#include "Utils/SpatialDebugger.h"
+
+#include "Debug/DebugDrawService.h"
+#include "Engine/Engine.h"
+#include "EngineClasses/SpatialNetDriver.h"
+#include "GameFramework/Pawn.h"
+#include "GameFramework/PlayerController.h"
+#include "GameFramework/PlayerState.h"
+#include "Interop/SpatialReceiver.h"
+#include "Interop/SpatialStaticComponentView.h"
+#include "Kismet/GameplayStatics.h"
+#include "Schema/AuthorityIntent.h"
+
+using namespace SpatialGDK;
+
+DEFINE_LOG_CATEGORY(LogSpatialDebugger);
+
+ASpatialDebugger::ASpatialDebugger(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+{
+	PrimaryActorTick.bCanEverTick = true;
+	PrimaryActorTick.bStartWithTickEnabled = true;
+
+	bAlwaysRelevant = true;
+	bNetLoadOnClient = false;
+	bReplicates = true;
+
+	NetUpdateFrequency = 1.f;
+
+	NetDriver = Cast<USpatialNetDriver>(GetNetDriver());
+
+	// For GDK design reasons, this is the approach chosen to get a pointer
+	// on the net driver to the client ASpatialDebugger.  Various alternatives
+	// were considered and this is the best of a bad bunch.
+	if (NetDriver &&
+		NetDriver->IsServer() == false)
+	{
+		NetDriver->SetSpatialDebugger(this);
+	}
+}
+
+void ASpatialDebugger::Tick(float DeltaSeconds)
+{
+	Super::Tick(DeltaSeconds);
+
+	check(NetDriver != nullptr);
+
+	if (NetDriver->IsServer() == false)
+	{
+		// Since we have no guarantee on the order we'll receive the PC/Pawn/PlayerState
+		// over the wire, we check here once per tick (currently 1 Hz tick rate) to setup our local pointers.
+		// Note that we can capture the PC in OnEntityAdded() since we know we will only receive one of those.
+		if (LocalPawn.IsValid() == false && LocalPlayerController.IsValid())
+		{
+			LocalPawn = LocalPlayerController->GetPawn();
+		}
+
+		if (LocalPlayerState.IsValid() == false && LocalPawn.IsValid())
+		{
+			LocalPlayerState = LocalPawn->GetPlayerState();
+		}
+
+		if (LocalPawn.IsValid())
+		{
+			SCOPE_CYCLE_COUNTER(STAT_SortingActors);
+			const FVector PlayerLocation = LocalPawn->GetActorLocation();
+
+			EntityActorMapping.ValueSort([PlayerLocation](const TWeakObjectPtr<AActor>& A, const TWeakObjectPtr<AActor>& B) {
+				return FVector::Dist(PlayerLocation, A->GetActorLocation()) > FVector::Dist(PlayerLocation, B->GetActorLocation());
+			});
+		}
+	}
+}
+
+void ASpatialDebugger::BeginPlay()
+{
+	Super::BeginPlay();
+
+	check(NetDriver != nullptr);
+
+	if (NetDriver->IsServer() == false)
+	{
+		EntityActorMapping.Reserve(ENTITY_ACTOR_MAP_RESERVATION_COUNT);
+
+		LoadIcons();
+
+		TArray<Worker_EntityId_Key> EntityIds;
+		NetDriver->StaticComponentView->GetEntityIds(EntityIds);
+
+		// Capture any entities that are already present on this client (ie they came over the wire before the SpatialDebugger did)
+		for (const Worker_EntityId_Key EntityId : EntityIds)
+		{
+			OnEntityAdded(EntityId);
+		}
+
+		// Register callbacks to get notified of all future entity arrivals / deletes
```

Comment:
MinorNit: General style is to have fullstops at the end of a comment.

Reply:
will fix